src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -140,38 +140,64 @@ internal void Freeze()
+        /// The alignment value is the necessary alignment for addresses for the underlying element type of the array.
+        /// The data is stored by using a type whose size is equal to the total size of the blob. When the total size
+        /// of the data is 1, 2, 4, or 8 bytes, a byte, short, int, or long can be used as the type of the field, and
+        /// regardless of what the element type is for the data, it's guaranteed to have appropriate alignment. For
+        /// all other sizes, a type is generated of the same size as the data, and that type needs its .pack set to
+        /// the alignment required for the underlying data. While that .pack value isn't required by anything else
+        /// in the compiler (the compiler always aligns RVA fields at 8-byte boundaries, which accomodates any
+        /// element type that's relevant), it is necessary for IL rewriters. Such rewriters also need to ensure
+        /// an appropriate alignment is maintained for the RVA field, and while they could also simplify by choosing
+        /// a worst-case alignment as does the compiler, they may instead use the .pack value as the alignment
+        /// to use for that field, since it's an opaque blob with no other indication as to what kind of data is
+        /// stored and what alignment might be required.
+        /// </param>
+        /// <returns>The field. This may have been newly created or may be an existing field previously created for the same data and alignment.</returns>
+        internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort alignment)
         {
-            switch (size)
-            {
-                case 1:
-                    return _systemInt8Type ?? new ExplicitSizeStruct(1, this, _systemValueType);
-                case 2:
-                    return _systemInt16Type ?? new ExplicitSizeStruct(2, this, _systemValueType);
-                case 4:
-                    return _systemInt32Type ?? new ExplicitSizeStruct(4, this, _systemValueType);
-                case 8:
-                    return _systemInt64Type ?? new ExplicitSizeStruct(8, this, _systemValueType);
-            }
+            Debug.Assert(!IsFrozen);
 
-            return new ExplicitSizeStruct(size, this, _systemValueType);
+            Cci.ITypeReference type = _proxyTypes.GetOrAdd(
+                ((uint)data.Length, Alignment: alignment), key =>
+                key.Size switch
+                {
+                    1 when _systemInt8Type is not null => _systemInt8Type,
+                    2 when _systemInt16Type is not null => _systemInt16Type,
+                    4 when _systemInt32Type is not null => _systemInt32Type,
+                    8 when _systemInt64Type is not null => _systemInt64Type,
+                    _ => new ExplicitSizeStruct(key.Size, key.Alignment, this, _systemValueType)
+                });
+
+            return _mappedFields.GetOrAdd((data, alignment, SpecialType.None), key =>
```

Comment:
>SpecialType.None [](http://example.com/codeflow?start=60&length=16)

I am not comfortable with usage of dummy values just so we could share a dictionary.